File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example3/SupportTicket.cs
namespace ChainOfResponsibilityDesignPattern.Example3;

public enum TicketPriority
{
    Low,
    Medium,
    High,
    Critical
}

public enum TicketCategory
{
    Technical,
    Billing,
    Account,
    General
}

public class SupportTicket
{
    public int TicketId { get; set; }
    public string CustomerName { get; set; }
    public string Issue { get; set; }
    public TicketPriority Priority { get; set; }
    public TicketCategory Category { get; set; }
    public bool IsResolved { get; set; }
    public string? ResolvedBy { get; set; }
    public string? Resolution { get; set; }

    public SupportTicket(int ticketId, string customerName, string issue, TicketPriority priority, TicketCategory category)
    {
        TicketId = ticketId;
        CustomerName = customerName;
        Issue = issue;
        Priority = priority;
        Category = category;
        IsResolved = false;
    }

    public void Resolve(string handler, string resolution)
    {
        IsResolved = true;
        ResolvedBy = handler;
        Resolution = resolution;
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example3/src/SupportTicket.ts
export enum TicketPriority {
  Low      = 0,
  Medium   = 1,
  High     = 2,
  Critical = 3,
}

export enum TicketCategory {
  Technical = "Technical",
  Billing   = "Billing",
  Account   = "Account",
  General   = "General",
}

export class SupportTicket {
  isResolved  = false;
  resolvedBy?: string;
  resolution?: string;

  constructor(
    public readonly ticketId: number,
    public readonly customerName: string,
    public readonly issue: string,
    public readonly priority: TicketPriority,
    public readonly category: TicketCategory
  ) {}

  resolve(handler: string, resolution: string): void {
    this.isResolved = true;
    this.resolvedBy = handler;
    this.resolution = resolution;
  }
}

export interface ISupportHandler {
  setNext(handler: ISupportHandler): ISupportHandler;
  handleTicket(ticket: SupportTicket): void;
}
